### PR TITLE
fix dotnet/templating #2457: 

### PR DESCRIPTION
**Pull request by vlada-shubina**
fix dotnet/templating #2457: 
Previously, in case running `dotnet new {T} --list --type {X}` and no exact matches by name and type the separate error message was shown for each template that has a match on name, but type is wrong. This could result in bulk of same text showing in differing only by template name.
Now instead only one error message is shown indicating number of template matched.

Note: the error is shown only when there is no exact matches by name and type. If there is at least one match, the list of matched templates are shown.

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs [hunk 1]
```diff
@@ -323,28 +323,33 @@ private static void DisplayPartialNameMatchLanguageAndContextProblems(string tem
...
-                    if ((matchInfo?.Kind ?? MatchKind.Mismatch) == MatchKind.Mismatch)
+                    // all templates in a group should have the same context & name
+                    if (templateGroup[0].Info.Tags != null && templateGroup[0].Info.Tags.TryGetValue("type", out ICacheTag typeTag))
                     {
-                        // {0} matches the specified name, but has been excluded by the --type parameter. Remove or change the --type parameter to use that template
-                        Reporter.Error.WriteLine(string.Format(LocalizableStrings.TemplateNotValidGivenTheSpecifiedFilter, templateGroup[0].Info.Name).Bold().Red());
+                        MatchInfo? matchInfo = WellKnownSearchFilters.ContextFilter(context)(templateGroup[0].Info);
+                        return ((matchInfo?.Kind ?? MatchKind.Mismatch) == MatchKind.Mismatch);
+                    }
+                    else
```
[the remark this answers is not in the document]
Comment (AR-May) [reply src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs:334]: This string seems indeed nonsense. I suppose we should change it. However, the comment before the string states this is not a real case. I have not seen this message on my runs of dotnet new3 also, but my experience is quite limited.

File: src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs [hunk 2]
```diff
@@ -323,28 +323,33 @@ private static void DisplayPartialNameMatchLanguageAndContextProblems(string tem
...
+                Reporter.Error.WriteLine(string.Format(LocalizableStrings.TemplatesNotValidGivenTheSpecifiedFilter, partialTemplatesMatchCount).Bold().Red());
+                anythingReported = true;
+            }
+            else if (partialTemplatesMatchCount == 1)
+            {
+                Reporter.Error.WriteLine(string.Format(LocalizableStrings.TemplateNotValidGivenTheSpecifiedFilter, partialTemplatesMatchCount).Bold().Red());
+                anythingReported = true;
             }
 
             if (templateResolutionResult.RemainingPartialMatchGroups.Count > 0)
```
[the remark this answers is not in the document]
Comment (AR-May) [reply src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs:355]: Yes, try to use, for example, -lang option for listing and not to use --type option. 
I got some items in RemainingPartialMatchGroups under these conditions.

[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs:355]: @AR-May thank you. That worked. 
However if you only use language without the type `DisplayPartialNameMatchLanguageAndContextProblems` runs but always returns on this code:
```
if (templateResolutionResult.IsNoTemplatesMatchedState || templateResolutionResult.UsingPartialMatches)
{
                ShowNoTemplatesFoundMessage(templateName, templateLanguage, context);
                return;
}
``` 
on second condition. 
Therefore the behavior for --lang filter is different from type.  
- `dotnet .\dotnet-new3.dll conf --debug:attach  --list  --type "project" -lang "C#"` - mismatch on the `type `- error is `1 template partially matched, but has been excluded by the --type parameter. Remove or change the --type parameter to use that template.`
- `dotnet .\dotnet-new3.dll conf --debug:attach  --list  --type "item" -lang "F#"1` - mismatch on the `language` - error is `No templates found matching: 'conf', lang=F#, type=item.`
- `dotnet .\dotnet-new3.dll conf --debug:attach  --list  -lang "F#" ` - mismatch on the `language`, no type error is `No templates found matching: 'conf', lang=F#, type=item.`
- dotnet .\dotnet-new3.dll conf --debug:attach  --list  -lang "F#" --type "project" - mismatch on the `language` and on `type` - error is `No templates found matching: 'conf', lang=F#, type=item.`

I can fix the code to apply the same behavior for language if needed but let's do it in another PR.

File: src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs:334]: @AR-May let's keep this code for now here, but also can we please discuss with Kathleen on how to adapt it, cause this string is only used in this place.
Current text is:  `{template name} cannot be created in the target location`
Suggested text is: `Cannot retrieve the type for {template name}` or `Cannot retrieve the context for {template name}`

File: src/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx [hunk 3]
```diff
@@ -347,7 +347,7 @@ See https://aka.ms/dotnet-install-templates to learn how to install additional t
     <value>The following templates partially match the input. Be more specific with the template name and/or language.</value>
   </data>
   <data name="TemplateNotValidGivenTheSpecifiedFilter" xml:space="preserve">
-    <value>{0} matches the specified name, but has been excluded by the --type parameter. Remove or change the --type parameter to use that template.</value>
+    <value>{0} template partially matched, but has been excluded by the --type parameter. Remove or change the --type parameter to use that template.</value>
```
Comment (bozturkMSFT) [inline src/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx:350]: Do we need the arg `{0}` since we already know that there is just "one" template here?

Or do we really need two strings here or can we just write `{0} template(s) partially matched,...` ? How is this situation handled in the rest of the localized strings?

File: src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs (hunk 1 above)
Comment (bozturkMSFT) [inline src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs:334]: Nit: since `if` above returns, we don't need the `else` here (we still need the code inside `else).

File: src/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx (hunk 3 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx:350]: @bozturkMSFT  I already asked this question before: we don't have any strings with "have/has" or any verbs with changing forms.  For plurals several strings just have (s). I think it makes sense to ask Kathleen probably to rephrase it to avoid have/has or use have(has).

File: src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (AR-May) [reply src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs:353]: Yes, indeed, the behavior for type and language options is quite different. Actually I'd like to unify the processing of type and language options in listing, but why not in another PR.

File: src/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx (hunk 3 above)
Comment (AR-May) [reply src/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx]: I suppose that one string is better than two. This string with have/has is the old one, just leave it be and add your own localized string. Kathleen already provided us the one without this problem: `12 template(s) partially matched, but failed on <whatever>`, where `<whatever>` should be `type=item`, for example.  `<whatever>` should be a parameter for the localized string, like in "No templates found matching" string. Later we will be able add support for the lang option there.

File: src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs [hunk 4]
```diff
@@ -323,28 +323,31 @@ private static void DisplayPartialNameMatchLanguageAndContextProblems(string tem
...
                     //{0} cannot be created in the target location
                     Reporter.Error.WriteLine(string.Format(LocalizableStrings.GenericPlaceholderTemplateContextError, templateGroup[0].Info.Name).Bold().Red());
                     anythingReported = true;
-                }
+                    return false;
+                });
+
+            if (partialTemplatesMatchCount > 1)
+            {
+                Reporter.Error.WriteLine(string.Format(LocalizableStrings.TemplatesNotValidGivenTheSpecifiedFilter, partialTemplatesMatchCount).Bold().Red());
```
Comment (AR-May) [inline src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs:344]: Let's use here the string Kathleen provided to us, not the old one: `12 templates partially matched, but failed on <whatever>`.

File: src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs]: Changed to: Cannot retrieve the type for {template name}

File: src/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx (hunk 3 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/LocalizableStrings.resx]: Changed to `{count} template(s) partially matched, but failed on {reason}`

File: src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs (hunk 4 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs]: Changed to `{count} template(s) partially matched, but failed on {reason}`

File: src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs:351]: The improved behavior will be implemented in https://github.com/dotnet/templating/issues/2472

**Review by bozturkMSFT — APPROVED** [review]

File: src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs [hunk 5]
```diff
@@ -319,32 +319,33 @@ private static void DisplayPartialNameMatchLanguageAndContextProblems(string tem
...
-            foreach (IReadOnlyList<ITemplateMatchInfo> templateGroup in templateResolutionResult.ContextProblemMatchGroups)
-            {
-                // all templates in a group should have the same context & name
-                if (templateGroup[0].Info.Tags != null && templateGroup[0].Info.Tags.TryGetValue("type", out ICacheTag typeTag))
+            int partialTemplatesMatchCount = templateResolutionResult.ContextProblemMatchGroups.Count(templateGroup =>
                 {
-                    MatchInfo? matchInfo = WellKnownSearchFilters.ContextFilter(context)(templateGroup[0].Info);
-                    if ((matchInfo?.Kind ?? MatchKind.Mismatch) == MatchKind.Mismatch)
+                    // all templates in a group should have the same context & name
+                    if (templateGroup[0].Info.Tags != null && templateGroup[0].Info.Tags.TryGetValue("type", out ICacheTag typeTag))
```
Comment (ladipro) [inline src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs:330]: nit: Since the `typeTag` argument is not used, this could be calling `ContainsKey()` instead of `TryGetValue()`.

File: src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs [hunk 6]
```diff
@@ -319,32 +319,33 @@ private static void DisplayPartialNameMatchLanguageAndContextProblems(string tem
...
                     anythingReported = true;
-                }
+                    return false;
+                });
+
+            if (partialTemplatesMatchCount > 0)
+            {
+                ShowNoTemplatesFoundMessage(templateName, templateLanguage, context);
+                // {0} template(s) partially matched, but failed on {1}.
+                Reporter.Error.WriteLine(string.Format(LocalizableStrings.TemplatesNotValidGivenTheSpecifiedFilter, partialTemplatesMatchCount, string.Concat("type=", context)).Bold().Red());
```
Comment (ladipro) [inline src/Microsoft.TemplateEngine.Cli/HelpAndUsage/HelpForTemplateResolution.cs:347]: nit/observation: There are many occurrences of the pattern `Reporter.Error.WriteLine(string.Format(...` in the code base so it would be worth adding an overload to `Microsoft.TemplateEngine.Cli` like so:

```C#
public void WriteLine(string format, params object[] arg)
```

(Just a random idea, out of the scope of this PR for sure.)

**Review by ladipro — APPROVED** [review]
The change looks great, thank you!

**Review by AR-May — APPROVED** [review]
